Fix error when 'Format-Wide -AutoSize | Out-String' is called.

stknohg opened the pull request:
## PR Summary

<!-- summarize your PR between here and the checklist -->
Fix #6471.

This pull request fixes the error when  `Format-Wide -AutoSize | Out-String` is called.  

From #5193,  `int.MaxValue` is adopted as output width when `Out-String -Width` parameter is not specified.  
`OutCommandInner.TableOutputContext` class is checking output width at initializing.

https://github.com/PowerShell/PowerShell/blob/36b71ba39e36be3b86854b3551ef9f8e2a1de5cc/src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs#L923-L939

This pull request applies the same method to `OutCommandInner.WideOutputContext` class.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link: #6471
- **Testing - New and feature**
    - [] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

stknohg committed:
Add checking if an output width is spcified.

stknohg committed:
add test.

iSazonov commented:
Do this fix  #6482 (PR #6487) too?

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Wide.Tests.ps1:
```diff
@@ -15,6 +15,8 @@ Describe "Format-Wide" -Tags "CI" {
 
     It "Should be able to use the autosize switch" {
         { Get-ChildItem | Format-Wide -Autosize } | Should -Not -Throw
+        # Test for #6471
```
Please add more useful comment why we should keep next test line.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Wide.Tests.ps1:
```diff
@@ -15,6 +15,8 @@ Describe "Format-Wide" -Tags "CI" {
     It "Should be able to use the autosize switch" {
         { Get-ChildItem | Format-Wide -Autosize } | Should -Not -Throw
+        # Test for #6471
+        { Get-ChildItem | Format-Wide -Autosize | Out-String } | Should -Not -Throw
```
What is current directory empty - do we get an exception without the fix?

stknohg replied:
Other tests are also invoked on the premise that there are files in the folder.
I think that it is probably okay.

iSazonov replied:
/cc @adityapatwardhan @JamesWTruher Can you approve the test?

SteveL-MSFT replied:
I think we should make this more predictable and consistent.  How about `Get-ChildItem $TestDrive` (and create at least one file there).  And since the test is for formatting and not testing `Get-ChildItem`, I would suggest doing that call once in `BeforeAll` and saving it to be used in the other tests.

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -1132,10 +1132,23 @@ internal override void Initialize()
+                    {
+                        columnsOnTheScreen = Console.WindowWidth;
+                    }
+                    catch
```
`int.MaxValue` is default - can we exclude the exception?

stknohg replied:
`OutCommandInner.TableOutputContext` class uses ` try-catch` statement assuming an environment without a console window such as implicit remoting.
Therefore, I do the same thing with `OutCommandInner.WideOutputContext` class.

>  // Tables need to use spaces for padding to maintain table look even if console window is resized. 
> // For all other output, we use int.MaxValue if the user didn't explicitly specify a width. 
>  // If we detect that int.MaxValue is used, first we try to get the current console window width. 
>  // However, if we can't read that (for example, implicit remoting has no console window), we default 
>  // to something reasonable: 120 columns.

iSazonov replied:
We already have reports of poor performance of the console. It is too costly to handle the exception for each row. My question is whether it is possible to exclude it.

SteveL-MSFT replied:
Since there are now at least two cases of this usage.  Perhaps we can have a static function get the windowwidth and if it caught an exception once, then have it return the default each time.

stknohg commented:
@iSazonov This fix only #6471.

stknohg committed:
add TryGetColumnsOnTheScreen method.

stknohg committed:
fix test comment.

stknohg committed:
change to use $TestDrive.

stknohg committed:
update code formatting(tab to space etc.).

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -743,6 +743,27 @@ private void ProcessCachedGroupOnWide(WideViewHeaderInfo wvhi, List<PacketInfoDa
             _formattingHint = hint;
         }
 
+        static private int TryGetColumnsOnTheScreen(int columnNumber)
```
Usually `Try...()` fuctions return `bool`. 
Also we get mainly `Console.WindowWidth`. So I think we could use `GetConsoleWindowWidth` name.

stknohg replied:
That's right. I will rename `TryGetColumnsOnTheScreen` to `GetConsoleWindowWidth`.

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -743,6 +743,27 @@ private void ProcessCachedGroupOnWide(WideViewHeaderInfo wvhi, List<PacketInfoDa
+            // If we detect that int.MaxValue is used, first we try to get the current console window width.
+            // However, if we can't read that (for example, implicit remoting has no console window), we default
+            // to something reasonable: 120 columns.
+            if (columnNumber == int.MaxValue)
```
Seems the comment is more for the method then the `if`. I believe we sgould move the comment.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Wide.Tests.ps1:
```diff
@@ -1,35 +1,40 @@
 Describe "Format-Wide" -Tags "CI" {
+    BeforeAll {
+        1..10 | ForEach-Object { New-Item -Path ("TestDrive:\Testdir{0:00}" -f $_) -ItemType Directory }
+        1..10 | ForEach-Object { New-Item -Path ("TestDrive:\TestFile{0:00}.txt" -f $_) -ItemType File }
```
10 is very large. 2 is enough.

stknohg replied:
I will fix it.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Wide.Tests.ps1:
```diff
@@ -1,35 +1,40 @@
-        $nonaliased = Get-ChildItem | Format-Wide
-        $aliased    = Get-ChildItem | fw
+        $nonaliased = Get-ChildItem $TestDrive | Format-Wide
+        $aliased = Get-ChildItem $TestDrive | fw
```
Please remove the alias and use full name.

My understanding @SteveL-MSFT request is that we should move `Get-ChildItem $TestDrive` to `BeforeAll` and reuse it in all `It` blocks:
```c#
$pathList = Get-ChildItem $TestDrive
...
$aliased = $pathList | Format-Wide 
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Wide.Tests.ps1:
```diff
@@ -1,35 +1,40 @@
-        { Get-ChildItem | Format-Wide -Autosize | Out-String } | Should -Not -Throw
+        { Get-ChildItem $TestDrive | Format-Wide -Autosize } | Should -Not -Throw
+        # 'Format-Wide -AutoSize | Out-String' fails on PowerShell Core 6.0.1. (issue #6471)
+        # so we add a new test for that.
```
Please remove unneeded comment. Each test is important.

stknohg replied:
I will remove it.

stknohg committed:
rename TryGetColumnsOnTheScreen to GetConsoleWindowWidth.

stknohg committed:
Fix review point.

* Change the count of files and directories used for testing from 10 to 2.
* Change to call 'Get-ChildItem' only once
* Remove unnecessary comment.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Wide.Tests.ps1:
```diff
@@ -1,33 +1,39 @@
-        $nonaliased = Get-ChildItem | Format-Wide
-        $aliased    = Get-ChildItem | fw
+        $nonaliased = $pathList | Format-Wide
+        $aliased = $pathList | fw
```
Please replace the alias.

stknohg replied:
What does `replace` mean?
This test compares the results of alias and unaliased command.

iSazonov replied:
Ah, sorry. We have special tests for aliases and shouldn't test it here. Pleas remove.

stknohg replied:
I understand. I will remove this test.

stknohg committed:
Remove unnecessary test.

SteveL-MSFT commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -743,6 +743,27 @@ private void ProcessCachedGroupOnWide(WideViewHeaderInfo wvhi, List<PacketInfoDa
+                }
+                catch
+                {
+                    return 120;
```
I think we can safely assert that if trying to get the WindowWidth ever fails, it will always fail.  So if we catch here, I think we can set a private flag like _failedToReadConsoleWidth and just return the default rather than incurring the cost of handling the exception each time (I think that was @iSazonov's original concern)

stknohg replied:
Does it really always fail?

I will present a simple example.   
I simply add `_failedToReadConsoleWidth` flag as follows.

```csharp
static private int _failedToReadConsoleWidth = 0;

static private int GetConsoleWindowWidth(int columnNumber)
{
    if (columnNumber == int.MaxValue)
    {
        // Tables and Wides need to use spaces for padding to maintain table look even if console window is resized.
        // For all other output, we use int.MaxValue if the user didn't explicitly specify a width.
        // If we detect that int.MaxValue is used, first we try to get the current console window width.
        // However, if we can't read that (for example, implicit remoting has no console window), we default
        // to something reasonable: 120 columns.
        if (_failedToReadConsoleWidth > 0)
        {
            return _failedToReadConsoleWidth;
        }
        try
        {
            return Console.WindowWidth;
        }
        catch
        {
            _failedToReadConsoleWidth = 120;
            return _failedToReadConsoleWidth;
        }
    }
    return columnNumber;
}
```

then invoke four commands in order.

* command1 : Invoke local command (Console.WindowWidth is available)
* command2 : Invoke implict remoting command (Console.WindowWidth isn't available)
* command3 : Invoke implict remoting command (Console.WindowWidth isn't available)
* command4 : Invoke local command (Console.WindowWidth is available)

value of `_failedToReadConsoleWidth` is changed to 120 after command2, and this value is used in command3.
But when we invoke command4, I think we have to use `Console.WindowWidth`.

iSazonov replied:
We should reset the flag.

stknohg replied:
Hmm...

In order to reset the flag, we need to be able to determine if `Console.WindowWidth` can get.  
if we can determine whether we can get `Console.WindowWidth` from the start, we do not need to use `try-catch` statement.

I am sorry but I have no idea that how to determine if `Console.WindowWidth` can get...

SteveL-MSFT replied:
In the implicit remoting case, the command would be executed on the remote side where it wouldn't be able to read the WindowWidth.  So the cached value would only be the remote side.  Local commands should still work.

stknohg replied:
@SteveL-MSFT Thank you. My premise was wrong.
I will fix the code as above.

stknohg committed:
Add '_failedToReadConsoleWidth' variable.

This variable is to cache the default console width when failed to get 'Console.WindowWidth' value.

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -743,6 +743,34 @@ private void ProcessCachedGroupOnWide(WideViewHeaderInfo wvhi, List<PacketInfoDa
+                // For all other output, we use int.MaxValue if the user didn't explicitly specify a width.
+                // If we detect that int.MaxValue is used, first we try to get the current console window width.
+                // However, if we can't read that (for example, implicit remoting has no console window), we default
+                // to something reasonable: 120 columns.
```
I'd move the comment before the method and add clarification about `_failedToReadConsoleWidth` and remote session.

stknohg replied:
OK.

However, I am not good at English and I do not have much confidence to make appropriate comments.
Is there any good comment plan?

iSazonov replied:
@SteveL-MSFT Could you please help with the comment?

SteveL-MSFT replied:
@stknohg Your English is actually quite good.  I believe @iSazonov is asking to move the comment to where line 748 right before the method.  You should probably just add some text before line 746 like:

```powershell
// In cases like implicit remoting, there is no console so reading the console width results in an exception.
// We cache this value rather than handling the exception (on the target) everytime as a performance improvement.
```

stknohg replied:
@SteveL-MSFT Thank you.I will add your comment.

SteveL-MSFT replied:
@stknohg I think this PR is almost there.  Just need to move the comment from lines 756-760 to right before line 750 since the comment applies to the entire method and not just the `if` statement within the method

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -743,6 +743,34 @@ private void ProcessCachedGroupOnWide(WideViewHeaderInfo wvhi, List<PacketInfoDa
+                // If we detect that int.MaxValue is used, first we try to get the current console window width.
+                // However, if we can't read that (for example, implicit remoting has no console window), we default
+                // to something reasonable: 120 columns.
+                if (_failedToReadConsoleWidth > 0)
```
I believe bool type is more suitable.
I'd rename this variable to reflect that a console is missing from the remote session.

SteveL-MSFT replied:
Something like `bool _noConsole`

stknohg committed:
Reflect review results.

* change _failedToReadConsoleWidth variable to _noConsole bool value.
* added comments.

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/BaseOutputtingCommand.cs:
```diff
@@ -743,6 +743,38 @@ private void ProcessCachedGroupOnWide(WideViewHeaderInfo wvhi, List<PacketInfoDa
+        // In cases like implicit remoting, there is no console so reading the console width results in an exception.
+        // We cache this value rather than handling the exception (on the target) everytime as a performance improvement.
+        static private bool _noConsole = false;
+
```
Sorry that I again feedback about comments.
We should follow the file pattern. Also I reword the comment.
```c#
/// <summary>
/// In cases like implicit remoting, there is no console so reading the console width results in an exception. 
/// Instead of handling exception every time we cache this value to increase performance. 
/// </summary>
static private bool _noConsole = false; 

/// <summary>
/// Tables and Wides need to use spaces for padding to maintain table look even if console window is resized.
/// For all other output, we use int.MaxValue if the user didn't explicitly specify a width.
/// If we detect that int.MaxValue is used, first we try to get the current console window width.
/// However, if we can't read that (for example, implicit remoting has no console window), we default
/// to something reasonable: 120 columns.
/// </summary>
static private int GetConsoleWindowWidth(int columnNumber)

stknohg replied:
No problem. I appreciate it.

stknohg committed:
Update comments to XML Documentation style.

SteveL-MSFT reviewed (approved):
Thanks for the contribution!

iSazonov commented:
@stknohg Thanks for the contribution!

Thank you for being patient - we want to have clean and clear code and helpful comments.
Now you have a useful experience and we would be happy to see your new contributions in formatting subsystem and other areas.
